Matchismo/Matchismo/Model/Card.h

```diff
@@ -14,7 +14,7 @@ struct MatchResult {
   NSMutableArray<Card *> *matches;
```

Comment:
Presumably this came from the provided code, but FYI using C-structs to hold Obj-C objects like this is pretty unusual, it's not something you'll see much in our codebase.

Reply:
I added that data structure myself to return two values from the function and be clear what types the two values are. Since objc doesn't have tuples is it normal to use an array or dictionary to return multiple values or is it done some other way? And how are different types of objects handled if an array is used, i.e. if the match function returned NSArray instead of MatchResult it isn't clear what the array contains so is it just assumed that the return type or what the array contains is known based on documentation and the objects are then cast to the correct type when used?

Reply:
It's true that there's no native concept of tuples in ObjC, though there are some third party libraries that provide them (in LT we often use RACTuple from the Reactive Cocoa library).
The alternative though is just to define a simple class with a couple of properties.
Usage of structs in ObjC isn't really equivalent to usage of structs in Swift.  In general they're just used when you're dealing with simple C-types (e.g. CGRect, CGSize and so on). 

I'd also add that it's generally a bad idea to have mutable fields on model objects, data containers and so on.  If these kinds of objects are entirely readonly (i.e. immutable) then it's much easier to reason about the code and avoid bugs arising from shared mutable state. 

So in this case, I guess you just define an interface something like this:
```
@interface MatchResult : NSObject

@property (readonly, nonatomic) NSInteger score;
@property (readonly, nonatomic) NSArray<Card *> *matches;

- (id)init NS_UNAVAILABLE;
- (instancetype)initWithScore:(NSInteger)score andMatches:(NSArray<Card *> *)matches NS_DESIGNATED_INITIALISER;

@end
```

Regarding mixed types in an array and casting the elements - you totally _could_ do this and objective c would allow it, but it's a bad idea as you can't leverage any safety checking from the compiler, so it would be really easy to make a mistake and crash the whole app.